File: Plantemon/app/src/main/java/com/g4ng/model/Player.java
package com.g4ng.model;

import java.util.List;

public class Player {
    // Conceptually, heals belong to the Player resource pool
    private static final int MAX_HEALS = 3;
    
    protected final String username;
    protected final List<Plant> garden;
    protected Plant currentPlant;

    private int remainingHeals = MAX_HEALS;

    public Player(String username, List<Plant> garden) {
        this.username = username;
        this.garden = garden;
    }

    public Plant getCurrentPlant() {
        return currentPlant;
    }

    public void setCurrentPlant(Plant currentPlant) {
        this.currentPlant = currentPlant;
    }

    public List<Plant> getGarden() {
        return garden;
    }

    public String getUsername() {
        return this.username;
    }

    public int getRemainingHeals(){
        return remainingHeals;
    }

    public void useHeal(){
        if (remainingHeals > 0) {
            remainingHeals--;
        }
    }

    public void resetHeals(){
        remainingHeals = MAX_HEALS;
    }

    public void restoreGarden() {
        resetHeals();
        for (Plant plant : garden) {
            plant.setCurrentHealth(plant.getMaxHealth());
        }
    }
}

File: plantemon-web/src/lib/domain/player.ts
import type { Plant } from "./plant";

const MAX_HEALS = 3;

/** Port of model/Player.java */
export class Player {
  readonly username: string;
  readonly garden: Plant[];

  private currentPlant: Plant | null = null;
  private remainingHeals = MAX_HEALS;

  constructor(username: string, garden: Plant[]) {
    this.username = username;
    this.garden = garden;
  }

  getUsername(): string {
    return this.username;
  }

  getGarden(): Plant[] {
    return this.garden;
  }

  getCurrentPlant(): Plant | null {
    return this.currentPlant;
  }

  setCurrentPlant(plant: Plant | null): void {
    this.currentPlant = plant;
  }

  getRemainingHeals(): number {
    return this.remainingHeals;
  }

  useHeal(): void {
    if (this.remainingHeals > 0) {
      this.remainingHeals--;
    }
  }

  resetHeals(): void {
    this.remainingHeals = MAX_HEALS;
  }

  restoreGarden(): void {
    this.resetHeals();
    for (const plant of this.garden) {
      plant.setCurrentHealth(plant.getMaxHealth());
    }
  }
}
